applications/IgaApplication/custom_utilities/director_utilities.cpp

```diff
@@ -42,9 +42,8 @@ namespace Kratos
             for (IndexType inodes = 0; inodes < number_of_control_points; ++inodes)
                 eqID[inodes] = r_geometry[inodes].GetId();
 
-            SparseMatrixType NTN(number_of_control_points, number_of_control_points, number_of_control_points*number_of_control_points); //inital guess how much non-zero are there
+            SparseMatrixType NTN = ZeroMatrix(number_of_control_points, number_of_control_points); //inital guess how much non-zero are there
```

Comment:
If this is not sparse any more the comment stops making sense.
This should be rolled-back to the sparse type. Additonally, the guess `number_of_control_points*number_of_control_points` is quite bad since it allocates the same size as a Dense Matrix. Maybe Kratos provides a way to get the connectiviy between notes to correctly set this allocation?

Reply:
Not sure if I am getting you right but in the way as it is now - it should be similarly efficient as the sparse matrix, no?

Reply:
The Sparse version was worse i think since it allocated the same memory? controlpoints^2 
and the sparse data access is slower.

Therefore, we have two option going back to sparse format with a better guess for the nonzeros or keep the dense version. with the dense version we are quite restricted since 10000controlpoints^2*8byte are already 4GB memory.

Reply:
@rath3t
I completely agree: The old sparse version is less efficient then the dense version.

Just keep in mind that 100000000 CPs for a single patch(!) is already a lot. At least, in practice, no one wants to work with such a CAD model anymore.

IMO it makes sense to merge the current implementation. It seems to work and you can already cover realistic CAD scenarios with it. An optimization can be done in the next step (if it is actually required).

Reply:
For me merging is also fine. But i want to point out that this is not an issue of a CAD model since also the model where the computation take place needs to call this method to get the reference nodal directors. Therefore, i think 10000CPs+ are not that far fetched. This would result in a stiffness matrix with 50000dofs which is not really big. But maybe i misunderstand what CAD model means.

Reply:
From what I understand, this method is called separately for each patch. Or am I wrong?

Reply:
Yes thats correct. You ask this because you assume that within one patch you don't have so much controlpoints, i suppose? This seems resonable, if this was your point.

Reply:
Exactly. From experience, complex NURBS-based CAD models tend to consist of many smaller patches.

But you are of course right that this improvement would make sense. I would just do it step by step with small PRs, so that there is always a working version in the main branch.
Next, for example, just re-change the matrix type to sparse and **omit** the initial-guess=#CPs^2. That might solve the memory problem.
In the next step you can think about how to initialize the structure of the matrix right at the beginning to enhance performance.